mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/transformation/mpsToModelix/initial/ModelSynchronizer.kt

```diff
@@ -123,83 +155,109 @@ class ModelSynchronizer(private val branch: IBranch, postponeReferenceResolution
 
-        // warning: might be fragile, because we synchronize the fields by hand
+        val targetModelReference = BuiltinLanguages.MPSRepositoryConcepts.ModelReference.model
         val targetModelModelixId = nodeMap[targetModel]!!
         val cloudTargetModel = branch.getNode(targetModelModelixId)
-        cloudModelReference.setReferenceTarget(
-            BuiltinLanguages.MPSRepositoryConcepts.ModelReference.model,
-            cloudTargetModel,
-        )
+        val idProperty = BuiltinLanguages.MPSRepositoryConcepts.Model.id
+        val cloudTargetModelId = cloudTargetModel.getPropertyValue(idProperty)
+
+        // duplicate check and sync
+        val modelImportExists = cloudParentNode.getChildren(childLink).firstOrNull {
+            cloudTargetModelId == it.getReferenceTarget(targetModelReference)?.getPropertyValue(idProperty)
+        } != null
+        if (modelImportExists) {
+            logger.warn { "Model Import for Model ${targetModel.name} from Model ${source.name} will not be synchronized, because it already exists on the server." }
+        } else {
+            val cloudModelReference =
+                cloudParentNode.addNewChild(childLink, -1, BuiltinLanguages.MPSRepositoryConcepts.ModelReference)
+
+            nodeMap.put(source, targetModel.reference, cloudModelReference.nodeIdAsLong())
+
+            // warning: might be fragile, because we synchronize the fields by hand
+            cloudModelReference.setReferenceTarget(targetModelReference, cloudTargetModel)
+        }
     }
 
     fun addLanguageDependency(model: SModel, language: SLanguage) =
         syncQueue.enqueue(linkedSetOf(SyncLock.MODELIX_WRITE, SyncLock.MPS_READ), SyncDirection.MPS_TO_MODELIX) {
             val modelixId = nodeMap[model]!!
-
-            val languageModuleReference = language.sourceModuleReference
+            val cloudNode = branch.getNode(modelixId)
             val childLink = BuiltinLanguages.MPSRepositoryConcepts.Model.usedLanguages
 
-            val cloudNode = branch.getNode(modelixId)
-            val cloudLanguageDependency =
-                cloudNode.addNewChild(
-                    childLink,
-                    -1,
-                    BuiltinLanguages.MPSRepositoryConcepts.SingleLanguageDependency,
+            val languageModuleReference = language.sourceModuleReference
+            val targetLanguageName = languageModuleReference?.moduleName
+            val targetLanguageId = languageModuleReference?.moduleId.toString()
+
+            // duplicate check and sync
+            val dependencyExists = cloudNode.getChildren(childLink)
+                .firstOrNull { targetLanguageId == it.getPropertyValue(BuiltinLanguages.MPSRepositoryConcepts.LanguageDependency.uuid) } != null
+            if (dependencyExists) {
+                logger.warn { "Model ${model.name}'s Language Dependency for $targetLanguageName will not be synchronized, because it already exists on the server." }
```

Comment:
tip: use early returns to avoid to much nesting.

Reply:
fixed in 54ed97610d6dc99daceab34dfa51119ad9c8cb56